### PR TITLE
Update LookupRow Indexes in Model when Settings Change

### LINKED ISSUE #33625: Update lookup-row indices in the model when settings have changed
- [x] When the Experiment Settings are changed, the `BatchPresenter::settingsChanged` method should call into the `Batch` model to perform updating of lookup row indices for all rows.
- [x] When a row's cell text has changed, the `RunsPresenter::updateRowField` is called. This should notify back up to the batch presenter that a row has changed. The batch presenter should then call the model to update the lookup row indices for that specific row. 
- [x] Similarly, `RowPresenter::updateGroupName` should notify back up to the batch presenter that a group has changed. The batch presenter should then call the model to update the lookup row indices for that specific group.

### PR DESCRIPTION
**Pull request by cailafinn**
**Description of work.**
Adds calls through the gui's presenter layers to inform the model that it needs to update rows. 
When:
- Exp Settings are changed, update all rows in the runs table
- a Group name is changed, update all rows in the group
- a Row's cell is changed, update that row

**To test:**
1. Load a batch, (ask me on slack for a file with some values already filled in, it's much easier than filling it all in yourself)
2. Take note of what matches should occur, (E.g: The run with an angle of 0.7 and a group name containing MAB should take the dQ/Q value from that row in the experiements settings tab)
3. Hit process on the runs tab
4. Check the right values were used. 

Fixes #33625 

*This does not require release notes* because **there are not yet any user facing changes.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by sf1919** [comment]
I had 2 files which you previously provided to me for testing #33688 . When I loaded either of these files in the ISIS Reflectometry GUI and then clicked 'Process' in the bottom right hand corner I got an error. Here is one of the stack traces as an example.

```
ReflectometryISISLoadAndProcess started
Error in execution of algorithm ReflectometryISISPreprocess:
Invalid value for property Filename (string) from string "13460": When setting value of property "Filename": File "13460" not found
  at line 56 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  caused by line 83 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
Error in execution of algorithm ReflectometryISISLoadAndProcess:
Invalid value for property Filename (string) from string "13460": When setting value of property "Filename": File "13460" not found
  at line 56 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  caused by line 83 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  at line 81 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 214 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 339 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
Error starting batch algorithm
Got error from algorithm "ReflectometryISISLoadAndProcess"
ReflectometryISISLoadAndProcess started
Error in execution of algorithm ReflectometryISISPreprocess:
Invalid value for property Filename (string) from string "13462": When setting value of property "Filename": File "13462" not found
  at line 56 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  caused by line 83 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
Error in execution of algorithm ReflectometryISISLoadAndProcess:
Invalid value for property Filename (string) from string "13462": When setting value of property "Filename": File "13462" not found
  at line 56 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  caused by line 83 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  at line 81 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 214 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 339 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
Error starting batch algorithm
Got error from algorithm "ReflectometryISISLoadAndProcess"
No matching experiment settings found for 13469. Using wildcard row settings instead. You may wish to check that all lookup criteria on the Experiment Settings table are correct.
ReflectometryISISLoadAndProcess started
Error in execution of algorithm ReflectometryISISPreprocess:
Invalid value for property Filename (string) from string "13469": When setting value of property "Filename": File "13469" not found
  at line 56 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  caused by line 83 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
Error in execution of algorithm ReflectometryISISLoadAndProcess:
Invalid value for property Filename (string) from string "13469": When setting value of property "Filename": File "13469" not found
  at line 56 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  caused by line 83 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  at line 81 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 214 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 339 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
Error starting batch algorithm
Got error from algorithm "ReflectometryISISLoadAndProcess"
ReflectometryISISLoadAndProcess started
Error in execution of algorithm ReflectometryISISPreprocess:
Invalid value for property Filename (string) from string "13470": When setting value of property "Filename": File "13470" not found
  at line 56 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  caused by line 83 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
Error in execution of algorithm ReflectometryISISLoadAndProcess:
Invalid value for property Filename (string) from string "13470": When setting value of property "Filename": File "13470" not found
  at line 56 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  caused by line 83 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISPreprocess.py'
  at line 81 in 'C:/..../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 214 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 339 in 'C:/.../mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
Error starting batch algorithm
Got error from algorithm "ReflectometryISISLoadAndProcess"
```

**Comment by sf1919** [comment]
> This looks like it's failing to find the run data. Is the archive switched on?

No - and thats fixed it. Need to remember to check that each time I change branch!

**Review by sf1919 — DISMISSED** [review]
I have undertaken a functional review and it works as expected. 

I have reviewed the code and now that my query has been answered I think it looks good and should function well.

File: qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp [hunk 1]
```diff
@@ -452,6 +452,7 @@ void RunsTablePresenter::updateGroupName(MantidWidgets::Batch::RowLocation const
     cell.setContentText(oldValue);
     m_view->jobs().setCellAt(itemIndex, column, cell);
   }
+  m_mainPresenter->notifyGroupNameChanged(m_model.mutableReductionJobs().mutableGroups()[0]);
```
Comment (gemmaguest) [inline qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp:455]: Should this use `groupIndex` rather than `0`?

File: qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp [hunk 2]
```diff
@@ -468,6 +469,7 @@ void RunsTablePresenter::updateRowField(MantidWidgets::Batch::RowLocation const
   updateRow(m_model.mutableReductionJobs(), groupIndex, rowIndex, rowValidationResult.validElseNone());
   if (rowValidationResult.isValid()) {
     showAllCellsOnRowAsValid(itemIndex);
+    m_mainPresenter->notifyRowContentChanged(m_model.mutableReductionJobs().mutableGroups()[0].mutableRows()[0].get());
```
Comment (gemmaguest) [inline qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp:472]: Should this use `groupIndex` and `rowIndex`?

File: qt/scientific_interfaces/ISISReflectometry/test/RunsTable/RunsTablePresenterDisplayTest.h [hunk 3]
```diff
@@ -245,6 +245,26 @@ class RunsTablePresenterDisplayTest : public CxxTest::TestSuite, RunsTablePresen
     verifyAndClearExpectations();
   }
 
+  void testNotifyBatchRowCellChanged() {
+    auto reductionJobs = twoGroupsWithMixedRowsModel();
+    auto presenter = makePresenter(m_view, reductionJobs);
+    auto const rowLocation = location(0, 0);
+    ON_CALL(m_jobs, cellAt(rowLocation, 0)).WillByDefault(Return(Cell("")));
+    EXPECT_CALL(m_mainPresenter, notifyRowContentChanged(_)).Times(1);
```
Comment (gemmaguest) [inline qt/scientific_interfaces/ISISReflectometry/test/RunsTable/RunsTablePresenterDisplayTest.h:253]: Could these tests check that the correct row/group is being passed in the expected call?

File: qt/scientific_interfaces/ISISReflectometry/test/RunsTable/RunsTablePresenterDisplayTest.h [hunk 4]
```diff
@@ -245,6 +245,26 @@ class RunsTablePresenterDisplayTest : public CxxTest::TestSuite, RunsTablePresen
     verifyAndClearExpectations();
   }
 
+  void testNotifyBatchRowCellChanged() {
+    auto reductionJobs = twoGroupsWithMixedRowsModel();
+    auto presenter = makePresenter(m_view, reductionJobs);
+    auto const rowLocation = location(0, 0);
```
Comment (gemmaguest) [inline qt/scientific_interfaces/ISISReflectometry/test/RunsTable/RunsTablePresenterDisplayTest.h:251]: Might be worth testing something other than index 0. Not sure it's worth a separate test but I tend to prefer to check the second group/row just because it's more likely to fail than finding the first one. It would also be clearer to create variables for the indices so it's clear how the expected outputs compare to the inputs (i.e. I'm not sure which `0` is which below, if that makes sense!)

File: qt/scientific_interfaces/ISISReflectometry/GUI/Save/SavePresenter.cpp [hunk 5]
```diff
@@ -244,12 +244,12 @@ void SavePresenter::saveWorkspaces(std::vector<std::string> const &workspaceName
 /** Saves selected workspaces */
 void SavePresenter::saveSelectedWorkspaces() {
   // Check that at least one workspace has been selected for saving
-  auto workspaceNames = m_view->getSelectedWorkspaces();
-  if (workspaceNames.empty()) {
+  auto wkspNames = m_view->getSelectedWorkspaces();
```
Comment (gemmaguest) [inline qt/scientific_interfaces/ISISReflectometry/GUI/Save/SavePresenter.cpp:247]: Was this intended to be changed? It's not a problem, just seems a bit unrelated to this PR?

**Review by gemmaguest — CHANGES_REQUESTED** [review]
Just a few queries related to what looks like potentially incorrect hard-coded indices and some possible improvements to the tests - the latter are not urgent for this PR if that causes problems for the subsequent PR

File: qt/scientific_interfaces/ISISReflectometry/GUI/Save/SavePresenter.cpp (hunk 5 above)
Comment (ConorMFinn) [reply qt/scientific_interfaces/ISISReflectometry/GUI/Save/SavePresenter.cpp:247]: This one was to fix a CppCheck failure that got flagged

File: qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp (hunk 1 above)
Comment (ConorMFinn) [reply qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/RunsTablePresenter.cpp:455]: Yes, it should. I think I picked that up in the next PR but forgot to come back and amend it here 🤦

File: qt/scientific_interfaces/ISISReflectometry/test/RunsTable/RunsTablePresenterDisplayTest.h (hunk 3 above)
Comment (ConorMFinn) [reply qt/scientific_interfaces/ISISReflectometry/test/RunsTable/RunsTablePresenterDisplayTest.h:253]: I can't really see a way to do this without some heavy refactoring. The updateRowField function makes a call to updateRow which causes the row in reduction jobs to be replaced. This makes it hard to compare the rows

**Comment by gemmaguest** [comment]
@sf1919 Could you check Conor's latest changes and re-approve if you're happy with them?

**Review by sf1919 — APPROVED** [review]
I have undertaken another functional test and it all works as expected. Looking over the code I have no further queries and am satisfied the changes requested to date have been implemented. I am therefore happy to approve.

**Review by gemmaguest — APPROVED** [review]
